assets/js/googlesitekit/datastore/user/utils.js

```diff
@@ -0,0 +1,104 @@
+
+const INVALID_DATE_INSTANCE_ERROR = 'Date param must construct to a valid date instance or be a valid date instance itself.';
+
+/**
+ * Asserts whether a given date instance is valid or invalid.
+ *
+ * @since n.e.x.t
+ *
+ * @param {Date} date Date instance to be asserted against.
+ * @return {boolean}  True if the given date instance is valid.
+ */
+export const isValidDateInstance = ( date ) => {
+	return date instanceof Date && ! isNaN( date );
+};
+
+/**
+ * Parses the given Date instance and returns a date string (YYYY-MM-DD).
+ *
+ * @since n.e.x.t
+ *
+ * @param {string|number|Date} date Date to parse into a string. Must be valid date value to be passed into Date constructor.
+ * @return {string}                 The parsed date string (YYYY-MM-DD).
+ */
+export const getDateString = ( date ) => {
+	const dateInstance = new Date( date );
+	invariant( isValidDateInstance( dateInstance ), INVALID_DATE_INSTANCE_ERROR );
+
+	const month = `${ dateInstance.getMonth() + 1 }`;
+	const day = `${ dateInstance.getDate() }`;
+	const year = dateInstance.getFullYear();
+
+	return [
+		year,
+		month.length < 2 ? `0${ month }` : month,
+		day.length < 2 ? `0${ day }` : day,
+	].join( '-' );
+};
+
+/**
+ * Parses the given date and returns the previous date (daysBefore).
+ *
+ * @since n.e.x.t
+ *
+ * @param {string} relativeDate Date string (YYYY-MM-DD) to subtract days from.
+ * @param {number} daysBefore   Number of days to subtract from relativeDate.
+ * @return {string}             The date string (YYYY-MM-DD) for the previous date.
+ */
+export const getPreviousDate = ( relativeDate = '', daysBefore ) => {
+	const dateArray = relativeDate.split( '-' );
+	const isValidRelativeDate = dateArray.length === 3 && isValidDateInstance( new Date( relativeDate ) );
+	invariant( isValidRelativeDate, INVALID_DATE_INSTANCE_ERROR );
+
+	const [ year, month, day ] = dateArray;
+	const date = new Date( year, month - 1, day );
+
+	date.setDate( date.getDate() - daysBefore );
+
+	return getDateString( date );
+};
+
+/**
+ * Similar to `getPreviousDate` but shifts the resulting date so that it returns a date
+ * that falls on the same day of the week as the `relativeDate` while preserving the
+ * `daysBefore` as a minimum.
+ *
+ * @since n.e.x.t
+ *
+ * @param {string} relativeDate Date string (YYYY-MM-DD) to subtract days from.
+ * @param {number} daysBefore   Number of days to subtract from relativeDate.
+ * @return {string}             The date string (YYYY-MM-DD) for the previous date, shifted to fall on the same day of the week.
+ */
+export const getPreviousWeekDate = ( relativeDate, daysBefore ) => {
+	const date = new Date( relativeDate );
+	invariant( isValidDateInstance( date ), INVALID_DATE_INSTANCE_ERROR );
+
+	const sameDay = daysBefore === 0;
+	const sameDayOfWeek = daysBefore % 7 === 0;
+	const addedDays = sameDay ? 7 : 7 - ( daysBefore % 7 );
+	const daysRemoved = daysBefore + ( sameDayOfWeek && ! sameDay ? 0 : addedDays );
```

Comment:
The PHP logic for this is slightly different as far as I can tell:
* I don't think we need to check for `sameDay`. If `daysBefore` is 0 (which is not a normal use-case for us), I'd expect this to return the exact same date as the input date. So that would also simplify this code here.
* This code currently will only add more days to `daysBefore`, but in the PHP logic it has some logic where it may add or remove days, based on what "same" weekday date is closer. Looking at this now, I'm not really sure whether your approach here or the one used in PHP is actually correct - any thoughts @adamsilverstein? As far as I can tell the current PHP logic may result in the date range including a few days that are also part of the other date range to compare with, which isn't something that I think would be intentional?

Reply:
> @adamsilverstein? As far as I can tell the current PHP logic may result in the date range including a few days that are also part of the other date range to compare with, which isn't something that I think would be intentional?

This is intentional and something I am pretty sure I saw in my testing, and what my internal ticket contacts told me to expect. It may not be expected (the whole aligning by day thing was surprising to me!), however I think this does match the current behavior of the Analytics dashboard.  If we are wrong it will show up in mismatched data (percent change) between Site Kit and Analytics.

The previous period range is adjusted to align with the last period range day of the week start, moving to the **closest** matching day. So if the last period starts on monday, the previous period should as well. if it starts on another day of the week, we shift to the closest monday - so if the day is tues, wed or thur, we move **back** to the previous monday - leaving a gap between the periods. if it is fri, sat or sunday, we move forward to the next monday - overlapping some days of the ranges. 

As it turns out, Site Kit's date ranges are all currently divisible by 7 - meaning the previous period will start on the same day - **except** the 90 day period which is off by one day. In this case, the logic will shift the previous period back one day, leaving a one day gap.

Reply:
This is very helpful insight on my end, I will edit the JS logic to move dates forward when `weekDayAlign` is `true` (when applicable) as they currently only move back in dates (this will cause date range overlap, but that seems to be expected).

Reply:
ps. Part of the reason the day alignment may make sense is that sites sometimes have distinct traffic patterns by day of the week (for example way more traffic/spikes on weekends). On a graph this will look like a series of humps, and when you display this against a previous period for comparison - you want those humps to line up in the graph!

Reply:
@adamsilverstein Another helpful comment, I appreciate it! I went ahead and did a bit more research within the original issue (when this was added within PHP) as well as testing this functionality within the native Analytics UI.

@felixarntz Changes to modify the above logic to match native Analytics UI compare/day align logic has been pushed up.